[docs] add first page for EAS builds (#9566)
expo/expo: docs/pages/build/eas-builds-in-5-minutes.md

---
title: EAS Builds in 5 Minutes
---

Let's get started by building Android and iOS app binaries for a fresh bare project initialized with Expo CLI.

## Prerequisites

> - Make sure you're using the latest version of Expo CLI. We're constantly improving the experience with EAS Builds, and since it's still in alpha, a slightly older version than the latest can become useless. **This tutorial assumes you're using `expo-cli@3.23.2` or newer.**

Reviewer: could say "Install `expo-cli@latest`" as one of the pre-reqs

## Create eas.json

The next step towards building your native project with EAS Builds is creating the `eas.json` file in the root directory. It will contain the EAS Builds configuration. Create the file with the following contents:

```json
{
  "builds": {
    "android": {
      "release": {
        "workflow": "generic"
      }
    },
    "ios": {
      "release": {
        "workflow": "generic"
      }
    }
  }
}
```

This declares that you want to build for Android and iOS. Each platform defines a single build profile named `release` (you can have multiple build profiles like `release`, `debug`, `testing`, etc.). Each profile declares that the project is a generic React Native project (unlike a managed Expo project which doesn't contain native code in the project tree).

After creating the file, make another commit:

- `git add eas.json`
- `git commit -m "Add eas.json"`

<center><img src="/static/images/eas-builds/5-minute-tutorial/04-eas-json.png" /></center>

If you want to learn more about `eas.json` see the [Configuring with eas.json](https://todo) page.

Once you've created `eas.json`, new Expo CLI commands should become available for you:

> - `expo eas:build` - manages the whole build process, which consists of: credentials management, project autoconfiguration, running build and waiting for it to finish

Reviewer: ```suggestion
- `expo eas:build` - manages the whole build process, which consists of: credentials management, project auto-configuration, running the build, and waiting for it to finish
```

auto-configuration meaning the changes to `build.gradle` (currently done manually) and `.pbxproject` files, right?

Author: > auto-configuration meaning the changes to build.gradle (currently done manually) and .pbxproject files, right?
yes :)

> - `expo eas:build:status` - displays the status of the lastest builds for the project

Reviewer: ```suggestion
- `expo eas:build:status` - displays the status of your latest build for this project
```

## Building For Android

> ### Configuring Android Project

Reviewer: ```suggestion
### Configuring the Android project
```

Author: Hmm, I'll add `the` but I'm not sure about lowercasing `Project`. I'm using this tool - https://capitalizemytitle.com/ and it says that `Project` should be capitalized.

Author: ```suggestion
### Configuring the Android Project
```

> We're working on streamlining this process. It'll look differently in the future. Stay tuned.

> So far, we've configured EAS Builds with `eas.json`. Now is the time to make the Android code buildable on our servers. Basically, the only thing you need to do is to provide your app's credentials to EAS Builds. As far as Android applications, you'll need a keystore, keystore password, key alias and key password to build the app binary.

Reviewer: ```suggestion
So far, we've configured EAS Builds with `eas.json`. Now is the time to make the Android code buildable on our servers. All you need to do is provide your app's credentials to EAS Builds. For an Android application, a keystore, keystore password, key alias, and key password are needed to build the app binary.
```

-            // see https://facebook.github.io/react-native/docs/signed-apk-android.
-            signingConfig signingConfigs.debug
+            signingConfig signingConfigs.release
             minifyEnabled enableProguardInReleaseBuilds
             proguardFiles getDefaultProguardFile("proguard-android.txt"), "proguard-rules.pro"
         }
```

We're defining here a new signing config (`release`) and using it for the `release` build type. The code in the `release` signing config is reading the keystore configration from the file named `credentials.json`. You don't need to create this file manually. It'll be automatically created on our build servers and populated with the credentials you generate (or provide) with Expo CLI. Using your own `credentials.json` is an advanced feature of EAS Builds. You can learn more about it in the [Advanced Credentials](https://todo) page.

After making the changes to the `build.gradle` file, commit them:

- `git add android/app/build.gradle`
- `git commit -m "Configure Android project"`

<center><img src="/static/images/eas-builds/5-minute-tutorial/05-configure-android.png" /></center>

### Running the Build

> Building the Android app binary is really straightforward. Just run `expo eas:build --platform android` and you'll be guided through generating or providing your own keystore. For the sake of simplicity of this tutorial, select `Generate new keystore` and hit `ENTER`.

Reviewer: ```suggestion
Building the Android app binary is simple- just run `expo eas:build --platform android` and you'll be guided through generating or providing your own keystore. For the sake of simplicity in this tutorial, select `Generate new keystore` and hit `ENTER`.
```

> Building the Android app binary is really straightforward. Just run `expo eas:build --platform android` and you'll be guided through generating or providing your own keystore. For the sake of simplicity of this tutorial, select `Generate new keystore` and hit `ENTER`.

Reviewer 2: We should add the info here that if the user already had credentials set up on app with the same slug it will try to use them

Author: Done.

<center><img src="/static/images/eas-builds/5-minute-tutorial/06-generate-keystore.png" /></center>

The build should start soon after that. Expo CLI will print a URL to the page where you can monitor the build (and view logs). Open the URL in a browser and you should see a page like this:

<center><img src="/static/images/eas-builds/5-minute-tutorial/07-build-progress.png" /></center>
<center><img src="/static/images/eas-builds/5-minute-tutorial/08-build-logs.png" /></center>

> Once the build completes you can download the app binary either by visiting the build status page (and clicking the download button) or by following the URL printed by Expo CLI:

Reviewer: ```suggestion
Once the build completes you can download the app binary by either visiting the build status page (and clicking the download button), or by opening the URL printed by Expo CLI:
```

<center><img src="/static/images/eas-builds/5-minute-tutorial/09-build-completed-website.png" /></center>

> <center><img src="/static/images/eas-builds/5-minute-tutorial/10-build-completed-cli.png" /></center>

Reviewer: would be good to highlight or circle the `.tar.gz` link

Open `app.json` and set the bundle identifier (`expo.ios.bundleIdentifer`) that will be used to identify your application on the Apple App Store.

An excerpt from Apple's docs:

> The bundle identifier string identifies your application to the system. This string must be a uniform type identifier (UTI) that contains only alphanumeric (A-Z,a-z,0-9), hyphen (-), and period (.) characters. The string should also be in reverse-DNS format. For example, if your company’s domain is Ajax.com and you create an application named Hello, you could assign the string com.Ajax.Hello as your application’s bundle identifier.

After setting the bundle identifier, commit the change:

- `git add app.json`
- `git commit -m "Set bundle identifier"`

<center><img src="/static/images/eas-builds/5-minute-tutorial/11-set-bundle-id.png" /></center>

### Running the Build

Run `expo eas:build --platform ios` and you'll be guided through generating the app's credentials and configuring your iOS project.

You'll be asked which bundle identifier to use. Choose the one defined in `app.json`:

<center><img src="/static/images/eas-builds/5-minute-tutorial/12-choose-bundle-id.png" /></center>

> Follow next steps (generating credentials and configuring the Xcode project) and you'll be provided with the URL to the build logs page:

Reviewer: ```suggestion
Follow the next steps (generating credentials and auto-configuring your Xcode project) and you'll be provided with the URL to the build logs page:
```

<center><img src="/static/images/eas-builds/5-minute-tutorial/13-ios-build-in-progress.png" /></center>

> Once the build completes you can download the app binary the same way of the Android counterpart - either by visiting the website or following the URL from console.

Reviewer: ```suggestion
Once the build completes you can download the app binary the same way you did with the Android counterpart - either via the build status page or opening the URL that's printed to your console.
```

---
title: EAS Builds in 5 Minutes
---

Let's get started by building Android and iOS app binaries for a fresh bare project initialized with Expo CLI.

## Prerequisites

> - Install Expo CLI by running `npm install -g expo-cli` (or `yarn global add expo-cli`). If you already have it, make sure you're using the latest version. We're constantly improving the experience with EAS Builds, and since it's still in alpha, a slightly older version than the latest can become useless. **This tutorial assumes you're using `expo-cli@3.23.2` or newer.**

Reviewer 3: ```suggestion
- Install Expo CLI by running `npm install -g expo-cli` (or `yarn global add expo-cli`). *If you already have it, make sure you're using the latest version.* EAS Builds is in alpha and it's changing rapidly, so the only way to ensure that you will have a good experience is to use the latest expo-cli version. **This tutorial assumes you're using `expo-cli@3.23.2` or newer.**
```

- Sign in with `expo login` or sign up with `expo register` if you don't have an Expo account yet. You can check if you're logged in by running `expo whoami`.

## Initialize a New Project

> Run `expo init PROJECT_NAME` (let's assume `PROJECT_NAME = abcd`) and choose a bare workflow template (either `minimal` or `minimal (TypeScript)`).

Reviewer 3: ```suggestion
Run `expo init PROJECT_NAME` (let's assume `PROJECT_NAME` is `abcd`) and choose a bare workflow template (either `minimal` or `minimal (TypeScript)`).
```

<center><img src="/static/images/eas-builds/5-minute-tutorial/01-init.png" /></center>

Initializing the project can take a few minutes. Once it's all set, you should see `✅ Your project is ready!` in logs.

<center><img src="/static/images/eas-builds/5-minute-tutorial/02-init-complete.png" /></center>

> Expo CLI creates a git repository for you. However, some changes are not commited by default. Go into the project directory, review the changes, and commit them:

Reviewer 3: we should fix this: https://github.com/expo/expo-cli/issues/2433

- `cd abcd`
- `git add .`
- `git commit -m "Add Podfile.lock, yarn.lock, etc."`

<center><img src="/static/images/eas-builds/5-minute-tutorial/03-initial-commit.png" /></center>

## Create eas.json

> The next step towards building your native project with EAS Builds is creating the `eas.json` file in the root directory. It will contain the EAS Builds configuration. Create the file with the following contents:

Reviewer 3: maybe we should add `expo eas:init` to create this file

Author: This sounds like a great idea! https://github.com/expo/turtle-v2/issues/272

```json
{
  "builds": {
    "android": {
      "release": {
        "workflow": "generic"
      }
    },
    "ios": {
      "release": {
        "workflow": "generic"
      }
    }
  }
}
```

This is a minimal example configuration for Android and iOS builds. It defines a single build profile named `release` (you can have multiple build profiles like `release`, `debug`, `testing`, etc.) for each platform. Each profile declares that the project is a generic React Native project (unlike a managed Expo project which doesn't contain native code in the project tree).

After creating the file, make another commit:

- `git add eas.json`
- `git commit -m "Add eas.json"`

<center><img src="/static/images/eas-builds/5-minute-tutorial/04-eas-json.png" /></center>

> If you want to learn more about `eas.json` see the [Configuring with eas.json](https://todo) page.

Reviewer 3: we also have a section in the API reference for configuration files now:

<img width="540" alt="Screen Shot 2020-08-06 at 10 13 42 AM" src="https://user-images.githubusercontent.com/90494/89561568-87830680-d7cd-11ea-991f-48853e704101.png">

https://docs.expo.io/versions/v38.0.0/config/app/

Author: 👍 I think that we should move that page there but only when we are ready for real users :)

**Warning: If you've previously set up credentials for an app with the same slug, Expo CLI will try to reuse them.**

<center><img src="/static/images/eas-builds/5-minute-tutorial/06-generate-keystore.png" /></center>

The build should start soon after that. Expo CLI will print a URL to the page where you can monitor the build (and view logs). Open the URL in a browser and you should see a page like this:

<center><img src="/static/images/eas-builds/5-minute-tutorial/07-build-progress.png" /></center>
<center><img src="/static/images/eas-builds/5-minute-tutorial/08-build-logs.png" /></center>

Once the build completes you can download the app binary by either visiting the build status page (and clicking the download button), or by opening the URL printed by Expo CLI:

<center><img src="/static/images/eas-builds/5-minute-tutorial/09-build-completed-website.png" /></center>
<center><img src="/static/images/eas-builds/5-minute-tutorial/10-build-completed-cli.png" /></center>

## Building for iOS

### Set the Bundle Identifier

> Open `app.json` and set the bundle identifier (`expo.ios.bundleIdentifer`) that will be used to identify your application on the Apple App Store.

Reviewer 3: i think we can leave this step out now that we read the bundle identifier from the xcode project right?

Author: Yes... and no :D Even though we read the bundle identifier from the Xcode project, the user has the ability to set his own id without opening Xcode - via app.json. As far as I remember, the default bundle identifier in the Xcode project generated with `expo init` is `org.name.PROJECT_NAME` and this isn't the greatest default that I can think of. If someone followed this tutorial and used the same name for the project (`abcd`) the credentials generation step would fail because I own the `org.name.abcd` app on the Apple App Store :)

Author: I'll describe this thoroughly in the `iOS` page.
